### PR TITLE
WIP: SymmetricKeyEncryptionManager

### PR DESCRIPTION
**Pull request by johnclara**
### Goal:
Share maintenance of logic in the community for doing symmetric key encryption.

Eventually this can be extended with specific dek providers (eg: aws specific implementations for aws KMS in iceberg-aws) and CryptoFormats (eg AES GCM/CTR similar to the s3 encrypted client from aws java sdk 1).

### Current Work
Based on Parquet encryption API here
https://github.com/apache/parquet-format/blob/master/Encryption.md

And implementation here:
https://issues.apache.org/jira/browse/PARQUET-1373

See SymmetricKeyEncryptionManager for the flows of reading and writing

### TODO/Open Questions
1. The current serde implementation is a flat json object. Future extensions could have a nested Json object or an avro object
2. Implement the CryptoFormat interface with AES encryption.
3. Perform formatting & add licenses
4. Pull into a core?

### How to Use:
1. Implement a key provider and a Serde (or reuse the json one)
example AWS java sdk2 implementation https://gist.github.com/johnclara/0ff9440d5d31392be47f6c0a19c06d00
2. Construct the provider/serde through Catalog properties in a custom catalog:
```
public class CustomCatalog extends AbstractCustomCatalog implements Configurable {
  private Configuration configuration;

  /**
   * Initialize and setConf are called by {@link org.apache.iceberg.CatalogUtil} during catalog creation through
   * some code paths (namely through the spark IcebergSource).
   * @param name TODO: do we need to propagate this down?
   * @param properties
   */
  @Override
  public void initialize(String name, Map<String, String> properties) {
    super.initialize(name, properties); // does nothing, but makes the linter happy
    initialize(properties, configuration);
  }
  
 @Override
  public void setConf(Configuration configuration) {
    this.configuration = configuration;
  }


  public void initialize(Map<String, String> properties, Configuration configuration) {
    Properties conf = Properties.of(properties);
    CustomCatalogConfig catalogConfig = CustomCatalogConfig.load(conf);
    initialize(catalogConfig, configuration);
  }

  public void initialize(CustomCatalogConfig catalogConfig, Configuration configuration) {
    AwsKmsKeyProvider kmsKeyProvider =
        new AwsKmsKeyProvider(
            catalogConfig.awsClientConfig(),
            catalogConfig.kmsKeyId(),
            catalogConfig.kmsClientFactoryImpl());

    Map<String, KeyProvider<? extends KeyMaterials>> keyProviders = new HashMap<>();
    keyProviders.put(kmsKeyProvider.id(), kmsKeyProvider);

    CryptoFormat cryptoFormat = new AesGcmFormat();

    Metastore metastore =
        new DynamoMetastore(
            catalogConfig.awsClientConfig(), catalogConfig.dynamoClientFactoryImpl());

    // Set custom config for hadoop at the normal namespace
    HadoopConf hadoopConf = HadoopConf.of(configuration);
    catalogConfig.dump(hadoopConf);

    HadoopConf baseConf = HadoopConf.of("", configuration);
    // Set s3a specific config at the base namespace
    catalogConfig.s3aConfig().setS3aHadoopConfig(baseConf);
    FileIO s3aFileIO = new HadoopFileIO(configuration);

    initialize(
        metastore,
        s3aFileIO,
        cryptoFormat,
        keyProviders,
        kmsKeyProvider
    );
  }
  ```
3. Create the SymmetricKeyCryptoManager and add it to table operations when operating with datafile Tables (it should not be added when interacting with other table types (like metadata tables) because it will try to decrypt the IO of the manifests which is broken)
```
public abstract class AbstractCustomCatalog extends BaseMetastoreCatalog {
  private Metastore metastore;
  private FileIO fileIO;
  private SymmetricKeyEncryptionManager encryptionManager;
  public AbstractCustomCatalog() {
  }

  public void initialize(
      Metastore metastore,
      FileIO fileIO,
      CryptoFormat cryptoFormat,
      Map<String, KeyProvider<? extends KeyMaterials>> keyProviders,
      KeyProvider<? extends KeyMaterials> keyProviderForNewKeys) {
    this.metastore = metastore;
    this.fileIO = fileIO;
    this.encryptionManager = new SymmetricKeyEncryptionManager(
        new KeyProviderDelegator(keyProviders, keyProviderForNewKeys),
        cryptoFormat,
        JsonKeyMetadataSerde.INSTANCE);
  }

  @Override
  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
    String tenantName = getTenantName(tableIdentifier);
    String tableName = getTableName(tableIdentifier);
    if (MetadataTableType.from(tableIdentifier.name()) != null) {
      return new AllTypesTableOperations(tenantName, tableName, metastore, fileIO);
    } else {
      return new DataTableOperations(tenantName, tableName, metastore, fileIO, encryptionManager);
    }
  }
...
```

4. Override the enryptionmanager in the DataTableOperations
```
public class DataTableOperations extends AllTypesTableOperations {
  private final SymmetricKeyEncryptionManager encryptionManager;

  public DataTableOperations(
      String tenantName,
      String icebergTableName,
      Metastore metastore,
      FileIO fileIO,
      SymmetricKeyEncryptionManager encryptionManager) {
    super(tenantName, icebergTableName, metastore, fileIO);
    this.encryptionManager = encryptionManager;
  }

  @Override
  public EncryptionManager encryption() {
    return encryptionManager;
  }
}
```

### CODE REVIEW

File: crypto/src/main/java/org/apache/iceberg/util/conf/NestedConf.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.util.conf;
+
+public class NestedConf extends Conf {
```
Comment (johnclara) [inline crypto/src/main/java/org/apache/iceberg/util/conf/NestedConf.java:22]: This is useful if you want to dump multiple KekIds into the properties when initializing the catalog.

example:
```
Properties props = Properties.of(new HashMap<>());
encryptedKekId.dump(NestedConf.of(props, "encrypted.kekid"));
plaintextKekId.dump(NestedConf.of(props, "plaintext.kekid"));
```

File: crypto/src/main/java/org/apache/iceberg/util/conf/JsonMapSerde.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import org.apache.iceberg.util.JsonUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JsonMapSerde implements MapSerde {
```
Comment (johnclara) [inline crypto/src/main/java/org/apache/iceberg/util/conf/JsonMapSerde.java:35]: This is legacy for my team. A different serde should probably be used

**Comment by jackye1995** [comment]
Thanks for sharing the patch! Let me take a close look this afternoon and see how this can integrate with the aws module.

**Comment by rdblue** [comment]
Thanks for working on this area, @johnclara!

Is your goal to just support a single encryption stream, or do you want to support Parquet and/or ORC encryption that still supports predicate pushdown?

**Comment by johnclara** [comment]
@jackye1995 might want to wait on trying to integrate. This is still really rough and I imagine the community will want to either change a bunch or start from scratch :) (or use @mccheah team's unreleased implementation instead)

**Review by jackye1995 — COMMENTED** [review]
Sorry for the delay, I was sick for the past few days.

First here are some general comments for styles (I decide to summarize here instead of commenting all over the place):
1. I see lots of terms like `dek`, `kek` directly used as variable names, I would suggest using the full name (at least in class names) for better context.
2. there are many code related to parsing and using configs, I think it is better to stick with the open source standards here and directly use `Map<String, String>` as config class and move all util classes to `PropertyUtil`.
3. I see multiple interfaces like `Dumpable` introduced but not used, which can be cleaned up a little bit.
4. there are comments to show the beginning of sections like private variables and constructors, which I don't think is needed in Iceberg standard.

For the general idea, I really like the idea of having a KEK + DEK architecture, I know multiple AWS customers using this architecture and it would be very helpful to have this feature in iceberg for wider adoption. 

For load a custom `DekProvider`, I think the traditional loading approach used in iceberg in multiple places should just work, why do you need another `DekProviders`? Can you provide more detail?

File: build.gradle [hunk 3]
```diff
@@ -1128,6 +1128,12 @@ project(':iceberg-nessie') {
 
 }
 
+project(':iceberg-crypto') {
```
[the remark this answers is not in the document]
Comment (johnclara) [reply build.gradle:1131]: For sure, I wanted to keep things separate while developing! I think I might split out a separate repo for testing addons that people could use as a template

**Comment by rdblue** [comment]
@johnclara, thanks for your work on this. I'm sorry that we haven't gotten back to the review yet. We're working on the 0.11.0 release. I think I should have time to revisit this after that.

**Comment by johnclara** [comment]
@rdblue starting to look at this again. will post a refactor with a lot changed soon

File: crypto/src/main/java/org/apache/iceberg/encryption/keys/KeyMetadata.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package org.apache.iceberg.encryption.keys;
+
+/**
+ * {@link KeyMetadata} metadata that is persistable into a byte array within a ManifestFile.
+ *
+ * <p>individual dekProviders are resposible for serializing and deserializing their
```
Comment (johnclara) [inline crypto/src/main/java/org/apache/iceberg/encryption/keys/KeyMetadata.java:6]: old naming. Should be KeyProvider and KeyMaterials

**Comment by RussellSpitzer** [comment]
cc @ggershinsky

**Comment by ggershinsky** [comment]
Hi guys, now that parquet-mr-1.12 is about to be released, we're starting to explore integration of Parquet column encryption in Iceberg. Thanks for looping me in, this pull request is certainly related. From the discussion above, and from the PR code, it looks like the focus is on general file encryption/decryption streams, and not on leveraging the native column encryption capabilities in Parquet and ORC. Which is fine, as these offer complementary sets of capabilities. Also, I agree these mechanisms share many common areas that we should define and re-use as much as possible. I'm sure this is possible; e.g. I see references to Parquet key management design and concepts like single and double wrapping. There is a number of open questions, though, like the storage of key material and an approach to key rotation; definition of key material, key metadata, key provider etc; protection of column statistics with global or column-specific keys; etc. I think it would be good to have a design document that draws a top-down picture of data encryption in Iceberg, with the goals and staging/roadmap; how this PR fits it, how ORC and Parquet column encryption will fit in later, what should be the common layers, etc.

**Comment by andersonm-ibm** [comment]
Hi @ggershinsky. I agree with you. I  think starting with a top-down design is a great idea and we would be happy to participate in the design and in the discussions around it.

**Comment by jackye1995** [comment]
@ggershinsky @andersonm-ibm 

I am in the process of exploring integration with Parquet and ORC column encryption to come up with a design that we can leverage format native encryption as much as possible while using the existing API and Iceberg architecture. There were also some good ideas for this during the secondary index discussion that I will incorporate here.

I am not sure if you were in the last sync meeting last week, as I said there, I am writing a design for this and will publish during the this weekend (was caught up with work and got delayed). I will send an email to dev list and update this PR when that is ready.

**Comment by shangxinli** [comment]
@ggershinsky @andersonm-ibm @jackye1995 
Great effort for designing encryption in Iceberg! I agree we should have a laying design and reuse as much as possible as both ORC and Parquet(1.12 is about to release) have column encryption.  Look forward to reviewing the design. 

There was also some discussion between ORC and Parquet communities on unifiying the table properties in Hive https://issues.apache.org/jira/browse/HIVE-21848.

**Comment by ggershinsky** [comment]
Cool, sounds good! 
@jackye1995 I too look forward to reviewing the design doc; planning this stuff top-down is the right thing to do. Btw, no need to rush it, we'll wait.
@shangxinli I agree it will be good to leverage the common column encryption configuration format we've defined in that jira.
@andersonm-ibm Glad you're joining this effort.

**Comment by jackye1995** [comment]
@ggershinsky @shangxinli @andersonm-ibm 
https://docs.google.com/document/d/1kkcjr9KrlB9QagRX3ToulG_Rf-65NMSlVANheDNzJq4/edit?usp=sharing

**Pull request closed without merging**